File: pkg/api/dtos/alertdefinition.go
package dtos

import (
	"time"
)

type AlertDefinition struct {
	Id        int64     `json:"id"`
	UpdatedAt time.Time `json:"updated_at"`

	Name           string                 `json:"name"`
	Description    string                 `json:"description"`
	AlertingMsg    string                 `json:"alerting_msg"`
	OkMsg          string                 `json:"ok_msg,omitempty"`
	AssociatedWith string                 `json:"associated_with"`
	Role           string                 `json:"role"`
	Severity       string                 `json:"severity"`
	For            *int64                 `json:"for,omitempty"`
	TicketEnabled  bool                   `json:"ticket_enabled,omitempty"`
	Configuration  map[string]interface{} `json:"configuration,omitempty"`
	Alerting       int64                  `json:"alerting"`
	Pending        int64                  `json:"pending"`
	Normal         int64                  `json:"normal"`
}

type AlertDefinitions struct {
	Count            int64             `json:"count"`
	Alerting         int64             `json:"alerting"`
	Pending          int64             `json:"pending"`
	Normal           int64             `json:"normal"`
	AlertDefinitions []AlertDefinition `json:"alert_definitions"`
	Page             int64             `json:"page"`
	PerPage          int64             `json:"perPage"`
}

type UpdateAlertDefinitionMsg struct {
	Id             int64                  `json:"id" binding:"required"`
	Name           string                 `json:"name" binding:"required"`
	Description    string                 `json:"description,omitempty"`
	AlertingMsg    string                 `json:"alerting_msg" binding:"required"`
	OkMsg          string                 `json:"ok_msg,omitempty"`
	AssociatedWith string                 `json:"associated_with" binding:"required"`
	Role           string                 `json:"role" binding:"required"`
	Severity       string                 `json:"severity" binding:"required"`
	For            *int64                 `json:"for,omitempty"`
	TicketEnabled  bool                   `json:"ticket_enabled"`
	Configuration  map[string]interface{} `json:"configuration,omitempty"`
}

type GetAlertDefinitionByIdMsg struct {
	Id int64 `json:"id"`
	// swagger:ignore
	Result AlertDefinition `json:"result"`
}

type SearchAlertDefinitionMsg struct {
	Query      string  `json:"query"`
	Page       int64   `json:"page"`
	PerPage    int64   `json:"perPage"`
	OrgId      int64   `json:"org_id"`
	ResourceId *int64  `json:"resource_id,omitempty"`
	GroupPath  *string `json:"group_path,omitempty"`
	State      *string `json:"state,omitempty"`
	User       User    `json:"user"`
	// swagger:ignore
	Result AlertDefinitions `json:"result"`
}

type IsAlertDefinitionAccessibleMsg struct {
	Id   int64 `json:"id" binding:"required"`
	User User  `json:"user"`
	// swagger:ignore
	Result bool `json:"result"`
}

File: public/app/types/devicemanagement/alert/alertdefinition.ts
import { OrgRole, SelectableValue } from '@grafana/data';
import { Configuration } from 'app/core/components/CustomForm/types';

export const alertDefinitionsPageLimit = 50;
export const severityTypes: Array<SelectableValue<string>> = [
  {
    label: 'Minor',
    value: 'minor',
    description: `Minor Alert`,
  },
  {
    label: 'Major',
    value: 'major',
    description: `Major Alert`,
  },
  {
    label: 'Critical',
    value: 'critical',
    description: `Critical Alert`,
  },
];

export interface TestAlertScriptDTO {
  measurement: string;
  uuid?: string;
  name?: string;
  resource_id?: number;
  group_path?: string;
  previous: any;
  data: any;
}

export const testAlertScriptDTO: TestAlertScriptDTO = {
  measurement: '',
  uuid: '',
  name: '',
  resource_id: 0,
  group_path: '0,',
  previous: {
    data1: '',
  },
  data: {
    data1: '',
  },
};

export interface AlertStats {
  count: number;
  alerting: number;
  pending: number;
  normal: number;
}

export interface AlertDefinition {
  id: number;
  updated_at: string;
  name: string;
  description: string;
  alerting_msg: string;
  ok_msg: string;
  associated_with: string;
  role: OrgRole;
  severity: string;
  for: number;
  ticket_enabled: boolean;
  configuration: Configuration;
  code: string;
  builtin: boolean;
  alerting: number;
  pending: number;
  normal: number;
  onEdit?: () => void;
}

export interface CreateAlertDefinitionDTO {
  name: string;
  description: string;
  alerting_msg: string;
  ok_msg: string;
  associated_with: string;
  role: OrgRole;
  severity: string;
  for: number;
  ticket_enabled: boolean;
  configuration: Configuration;
  code: string;
}

export interface UpdateAlertDefinitionDTO {
  name: string;
  description: string;
  alerting_msg: string;
  ok_msg: string;
  associated_with: string;
  role: OrgRole;
  severity: string;
  for: number;
  ticket_enabled: boolean;
  configuration: Configuration;
  code: string;
}

export interface AlertDefinitionsState {
  alertDefinitions: AlertDefinition[];
  alertDefinitionsStats: AlertStats;
  searchQuery: string;
  searchPage: number;
  hasFetched: boolean;
}

export interface AlertDefinitionState {
  alertDefinition: AlertDefinition;
}

export interface AlertDefinitionDTO {
  page: number;
  association?: string;
  associationReference?: string | number;
  state?: string;
  query?: string;
}

export const defaultCode = `
//result object  expected as response 
//state = PENDING will skip alert execution
var result = { err: '', state: PENDING };

//uplink object recieved
var msg = getMessage();
if (msg.Error !== "") {
  result.err = msg.Error
  return result;
}

//alert instance for the uplink recieved
var alert = getAlert();
if (alert.Error !== "") {
  result.err = alert.Error
  return result;
}

//check for any configuration required for alert validation
if (!alert.Result.Configuration || !alert.Result.Configuration.thershold) {
  result.err = 'configuration not found';
  return result;
}
var thershold = alert.Result.Configuration.thershold;

//group filter from the uplink recieved
var grpFilter = getGroupFilter();
if (grpFilter.Error !== "") {
  result.err = grpFilter.Error;
  return result;
}

//Read data from influx if required
var query = "";

var readData = readRawData(query);
if (readData.Error !== "") {
  result.err = readData.Error;
  return result;
}

//validate weather result is valid
if (!readData.Result.hasOwnProperty('consumption')) {
  //Do nothing
  return result;
}

//get group from uplink recieved
var group = getGroup();
if (group.Error !== "") {
  result.err = group.Error;
  return result;
}

//check alert here
result.state = OK;
if (Number(readData.Result.consumption) > Number(thershold)) {
  result.state = ALERTING;
  setContext('consumption', readData.Result.consumption)
}
return result;
`;
